Adding `Subtensor` static shape inference in `Subtensor.make_node`

larryshamalama opened the pull request:
Closes #922.

As per @brandonwillard's [comment](https://github.com/aesara-devs/aesara/issues/922#issuecomment-1108901650), I renamed `broadcastable` to `static_shape`, incorporating the other suggested minor edits.

I took inspiration from [lines 773 to 777](https://github.com/aesara-devs/aesara/blob/main/aesara/tensor/subtensor.py#L773-L777) from `Subtensor.infer_shape` for shape inference.

No new tests have been added yet as I am creating this PR.

larryshamalama commented:
N.B.: Local testing yields `CompilationError`s because `mkl` provides a C++ interface via Python, except installation of the package is difficult on an M1 Mac (c.f. issue #909). Would need to think of a (possibly hackish) solution, at least for myself

brandonwillard reviewed (changes requested):
[The error in CI](https://github.com/aesara-devs/aesara/runs/6211102663?check_suite_focus=true#step:6:105) says that the rewrite `local_subtensor_make_vector` introduced a type error when it tried to replace a variable with another variable that has an incompatible static shape.

Inspection via the debugger shows that the node being rewritten has the following form:

> Subtensor{int64::} [id BH] <TensorType(int64, (5,))> '' 4
> 
> | MakeVector{dtype='int64'} [id BD] <TensorType(int64, (3,))> ''   3 |
> | ScalarConstant{-2} [id P] <int64>                                  |

A shape of `(5,)` for the output of a `Subtensor` `Op` like `x[-2::]`, where `x` is a vector of length 3, doesn't make much sense.

Let's see if we can reproduce the error by creating an equivalent graph.

```python
import aesara
import aesara.tensor as at


# A vector of length three built from a `MakeVector` `Op`
x = at.as_tensor([at.lscalar(), at.lscalar(), at.lscalar()])
z = x[-2:]

aesara.dprint(z, print_type=True, depth=2)
# Subtensor{int64::} [id A] <TensorType(int64, (5,))> ''
#  |MakeVector{dtype='int64'} [id B] <TensorType(int64, (3,))> ''
#  |ScalarConstant{-2} [id C] <int64>
```

We've successfully reproduced the problematic graph, so we can now isolate the issue in `Subtensor.make_node`.

Also, we now have a good unit test to add to `tests.tensor.test_subtensor`. Since all the current tests in `tests.tensor.test_subtensor` seem to pass when this bug is present, it's clear that those tests are lacking coverage for some simple and important cases. One of which is covered by the simple example above.

FYI: Simple test additions like these can be just as valuable to Aesara as the feature work from which they are derived.

larryshamalama commented:
Thanks for the detailed input. I see where the error is and will try to fix it. Maybe I can reuse `get_canonical_form_slice` in some way, but I'm working on it.

Right now I am spending time addressing issues with testing on my Mac. I was running into graph compilation issues in `pytest` but it turns out that they are due to the lack of `mkl`...
